src/streamlink/plugins/vimeo.py

```diff
@@ -8,70 +8,175 @@
-                validate.url(),
-            ),
-        ),
-    )
+class VimeoMuxedHLSStream(MuxedHLSStream):
+    def __init__(self, session, video, audio, **args):
+        super().__init__(session, video, audio, hlsstream=VimeoHLSStream, **args)
+
+
+class VimeoHLSStream(HLSStream):
+    __muxed__ = VimeoMuxedHLSStream
+
+    _expiry_time = 30
+    _expiry_re1 = re.compile(r"^exp=(\d+)~")
+    _expiry_re2 = re.compile(r"^(\d+)-")
+
+    def __init__(self, session_, url, **args):
+        self.session = session_
+        self._url = url
+        self._parsed_url = urlparse(self._url)
+        self._path_parts = self._parsed_url.path.split("/")
+        self._expiry = self.get_expiry(self._path_parts[1])
+        super().__init__(self.session, self._url, **args)
+
+    def _update_auth(self):
+        self.session.api._updating = True
+
+        data = Vimeo._get_data(self)
+        if not data:
+            raise ValueError("no video data found")
+        videos = data["request"]["files"]
+        if "hls" not in videos:
+            raise ValueError("hls key not found in video data")
+
+        new_auth = None
+        exit_loop = False
+        for _, video_data in videos["hls"]["cdns"].items():
+            url = video_data.get("url")
+            for _, stream in VimeoHLSStream.parse_variant_playlist(self.session, url).items():
+                parsed_url = urlparse(stream.to_manifest_url())
+                path_parts = parsed_url.path.split("/")
+                if self._expiry_re1.search(path_parts[1]) and self._expiry_re1.search(self._path_parts[1]):
+                    new_auth = path_parts[1]
+                elif self._expiry_re2.search(path_parts[1]) and self._expiry_re2.search(self._path_parts[1]):
+                    new_auth = path_parts[1]
+
+                if new_auth:
+                    self.session.api._auth_data = new_auth
+                    exit_loop = True
+                    break
+            if exit_loop:
+                break
+        if not new_auth:
+            raise ValueError("failed to get new auth data from URL")
+
+        self.session.api._last_updated = time()
+        self.session.api._updating = False
+
+    @staticmethod
+    def get_expiry(path):
+        m = VimeoHLSStream._expiry_re1.search(path) or VimeoHLSStream._expiry_re2.search(path)
+        if not m:
+            raise ValueError("expiry value not found in URL")
+        return int(m.group(1))
+
+    @property
+    def url(self):
+        time_now = time()
+        if time_now > self._expiry - self._expiry_time:
+            if self.session.api._updating:
+                for i in range(self._expiry_time):
+                    log.debug(f"Waiting on Vimeo auth data update: {i}s")
+                    sleep(1)
+                    if not self.session.api._updating:
+                        break
+                else:
+                    raise PluginError("timed out waiting for auth data update to finish")
```

Comment:
That seems super weird. Why do you need to wait for the API update? `api._updating` can only get set to `True` via `_update_auth()`, which is called from the same thread, namely the `HLSStreamWorker` thread. And `sleep()` should never be used when waiting for the availability of a resource, that's what thread locks are for. Raising a `PluginError` is also not handled by the `url` property.

Reply:
Yes, that's why.  Noted about `sleep()`.